### PR TITLE
Auto sync issues

### PR DESCRIPTION
**Pull request by damien-urruty-sonarsource**

### CODE REVIEW

File: src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  private Set<String> tryUpdateConnectionsAndBoundProjectStorages(Map<String, Map<String, Set<String>>> projectKeyByConnectionIdsToUpdate, ProgressFacade progress) {
+    var failedConnectionIds = new LinkedHashSet<String>();
+    projectKeyByConnectionIdsToUpdate.forEach(
+      (connectionId, projectKeys) -> tryUpdateConnectionAndBoundProjectsStorages(projectKeyByConnectionIdsToUpdate, progress, failedConnectionIds, connectionId, projectKeys));
+    return failedConnectionIds;
+  }
+
+  private void tryUpdateConnectionAndBoundProjectsStorages(Map<String, Map<String, Set<String>>> projectKeyByConnectionIdsToUpdate, ProgressFacade progress,
+    Set<String> failedConnectionIds, String connectionId, Map<String, Set<String>> branchNamesByProjectKey) {
+    progress.doInSubProgress(connectionId, 1.0f / projectKeyByConnectionIdsToUpdate.size(), subProgress -> {
```
Comment (sophio-japharidze-sonarsource) [inline src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java:109]: Just to verify, will `projectKeyByConnectionIdsToUpdate.size()` always be >0?

File: src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        branchNamesByProjectKey.keySet(), endpointParamsAndHttpClient, engineOpt.get(), s));
+      subProgress.doInSubProgress("Sync projects storages", 0.5f, s -> syncOneEngine(
+        connectionId, branchNamesByProjectKey, engineOpt.get(), s));
+    });
+  }
+
+  private static void tryUpdateBoundProjectsStorage(Set<String> projectKeys, ServerConnectionSettings.EndpointParamsAndHttpClient endpointParamsAndHttpClient,
+    ConnectedSonarLintEngine engine,
+    ProgressFacade progress) {
+    projectKeys.forEach(projectKey -> progress.doInSubProgress(projectKey, 1.0f / projectKey.length(), subProgress -> {
```
Comment (sophio-japharidze-sonarsource) [inline src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java:130]: I think here is a typo, should be `1.0f / projectKey**s**.length()`

File: src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  private void syncStorage() {
+    var projectKeyPerConnectionId = new HashMap<String, Set<String>>();
+    bindingManager.forEachBoundFolder((folder, settings) -> {
+      var connectionId = requireNonNull(settings.getConnectionId());
+      var projectKey = requireNonNull(settings.getProjectKey());
+      projectKeyPerConnectionId
+        .computeIfAbsent(connectionId, k -> new HashSet<>())
+        .add(projectKey);
+    });
+    if (!projectKeyPerConnectionId.isEmpty()) {
```
Comment (kirill-knize-sonarsource) [inline src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java]: Maybe flip the condition and return if empty?

File: src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java (hunk 1 above)
Comment (damien-urruty-sonarsource) [reply src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java:109]: Yes. I changed the code a bit to make it more obvious

File: src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java (hunk 2 above)
Comment (damien-urruty-sonarsource) [reply src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java:130]: Good catch!

File: src/test/java/org/sonarsource/sonarlint/ls/connected/ProjectBindingManagerTests.java [hunk 4]
```diff
@@ -796,12 +641,11 @@ private WorkspaceFolderWrapper mockFileInABoundWorkspaceFolder() {
     return folder;
   }
 
-  private WorkspaceFolderWrapper mockFileInABoundWorkspaceFolder2() {
+  private void mockFileInABoundWorkspaceFolder2() {
```
Comment (kirill-knize-sonarsource) [inline src/test/java/org/sonarsource/sonarlint/ls/connected/ProjectBindingManagerTests.java:644]: Methods with similar names works differently now. A bit confusing. Maybe it's not a bad thing to ignore return value.

File: src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java (hunk 3 above)
Comment (damien-urruty-sonarsource) [reply src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java]: The code has moved and the method is a bit shorter. I like to exit early if the method is long, here I think it's ok

File: src/test/java/org/sonarsource/sonarlint/ls/connected/ProjectBindingManagerTests.java (hunk 4 above)
Comment (damien-urruty-sonarsource) [reply src/test/java/org/sonarsource/sonarlint/ls/connected/ProjectBindingManagerTests.java:644]: OK I will keep it for consistency

**Review by kirill-knize-sonarsource — APPROVED** [review]

**Pull request merged**